src/rust/src/wallet_scanner.rs

```diff
@@ -0,0 +1,470 @@
+use std::collections::HashMap;
+use std::io;
+use std::mem;
+use std::sync::mpsc;
+
+use group::GroupEncoding;
+use zcash_note_encryption::Domain;
+use zcash_note_encryption::{batch, BatchDomain, ShieldedOutput, ENC_CIPHERTEXT_SIZE};
+use zcash_primitives::constants;
+use zcash_primitives::{
+    consensus,
+    sapling::{self, note_encryption::SaplingDomain},
+    transaction::{
+        components::{sapling::GrothProofBytes, OutputDescription},
+        Transaction, TxId,
+    },
+};
+
+#[cxx::bridge]
+mod ffi {
+    #[namespace = "wallet"]
+    struct DecryptionResult {
+        txid: [u8; 32],
+        output: u32,
+        ivk: [u8; 32],
+        diversifier: [u8; 11],
+        pk_d: [u8; 32],
+    }
+
+    #[namespace = "wallet"]
+    extern "Rust" {
+        type Network;
+        type BatchScanner;
+        type BatchResult;
+
+        fn network(
+            network: &str,
+            overwinter: i32,
+            sapling: i32,
+            blossom: i32,
+            heartwood: i32,
+            canopy: i32,
+            nu5: i32,
+        ) -> Result<Box<Network>>;
+
+        fn init_batch_scanner(
+            network: &Network,
+            sapling_ivks: &[[u8; 32]],
+        ) -> Result<Box<BatchScanner>>;
+        fn add_transaction(self: &mut BatchScanner, tx_bytes: &[u8], height: u32) -> Result<()>;
+        fn flush(self: &mut BatchScanner);
+        fn collect_results(self: &mut BatchScanner, txid: [u8; 32]) -> Box<BatchResult>;
+
+        fn get_sapling(self: &BatchResult) -> Vec<DecryptionResult>;
+    }
+}
+
+/// The minimum number of outputs to trial decrypt in a batch.
+///
+/// TODO: Tune this.
+const BATCH_SIZE_THRESHOLD: usize = 20;
+
+#[derive(Clone, Copy)]
+pub enum Network {
+    Consensus(consensus::Network),
+    RegTest {
+        overwinter: Option<consensus::BlockHeight>,
+        sapling: Option<consensus::BlockHeight>,
+        blossom: Option<consensus::BlockHeight>,
+        heartwood: Option<consensus::BlockHeight>,
+        canopy: Option<consensus::BlockHeight>,
+        nu5: Option<consensus::BlockHeight>,
+    },
+}
+
+/// Constructs a `Network` from the given network string.
+///
+/// The heights are only for constructing a regtest network, and are ignored otherwise.
+fn network(
+    network: &str,
+    overwinter: i32,
+    sapling: i32,
+    blossom: i32,
+    heartwood: i32,
+    canopy: i32,
+    nu5: i32,
+) -> Result<Box<Network>, &'static str> {
```

Comment:
That would require switching everywhere we want to do this (instead of only here) on the C++ side (where we don't have things as nice as the Rust `match`). I might have done this if I could hook it directly into the chainparams logic, but:
- I don't think the compilation units are set up to enable that at the moment, so it would be a more invasive change.
- `cxx` doesn't let you reuse FFI definitions yet, so it can't be a 1:1 mapping for now.